### PR TITLE
Write SOP (resolves #85)

### LINKED ISSUE #85: Add SOP for running RNAseq samples in production

### PR DESCRIPTION
**Pull request by jvivian**

### CODE REVIEW

**Comment by jvivian** [comment]
@hannes-ucsc — This is just a first stab. I figured I'd let @JakeNarkizian use this as his template for the next production run and he can give me feedback if anything is unclear.

File: Production_SOP.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+### Setup CGCloud
+Comprehensive help is found at: https://github.com/BD2KGenomics/cgcloud .
+ The following steps are provided for convenience but it is recommended to read the source documentation.
+
+1. `virtualenv ~/cgcloud`
+2. `source cgcloud/bin/activate`
+3. `pip install cgcloud-toil==1.5.5`
+4. `deactivate`
+5. `ln -snf ~/cgcloud/bin/activate ~/bin/`
+6. Set `~/.profile` options from CGCloud README
```
Comment (hannes-ucsc) [inline Production_SOP.md:20]: Be explicit. No linking or references except for optional steps. This document should be self contained. Be sure to mention `CGCLOUD_KEYPAIRS`.

File: Production_SOP.md [hunk 2]
```diff
@@ -0,0 +1,71 @@
+# Core Operations Production SOP
+## UC Santa Cruz Genomics Institute, Computational Genomics Lab
+ 
+## Environment Setup
+ 
+ These steps should only have to be run once. Toil-box images will have to be updated periodically to stay current.
```
Comment (hannes-ucsc) [inline Production_SOP.md]: So for readability you should always backquote strings that a user would type into a command line.

So for example:

```
Bash can be started by running the `bash` command.
```

Or in this case:

```
The `toil-box` image will …
```

Comment (hannes-ucsc) [reply Production_SOP.md]: "periodically" is imprecise. When exactly should they update?

File: Production_SOP.md (hunk 1 above)
Comment (hannes-ucsc) [reply Production_SOP.md:20]:
```
The following steps are provided for referecnce but it is recommended to read the source documentation to gain full understanding on how to use CGCloud.
```

File: Production_SOP.md [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  
+### Setup CGCloud
+Comprehensive help is found at: https://github.com/BD2KGenomics/cgcloud .
```
Comment (hannes-ucsc) [inline Production_SOP.md]: Markdown can do hyperlinks.

File: Production_SOP.md [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+### Launching cluster
+Launch the cluster leader first, and grow the cluster after starting the pipeline.
+The directory passed to `--share` should contain the master key if encryption is being used.
+
+- `cgcloud create-cluster --leader-instance-type m3.medium --share=key_dir/ --num-workers 0 --cluster-name example toil`
+
+### Setup and Run Pipeline
+SSH into leader, install toil-scripts, setup config and manifest, grow cluster, and start run.
+
+1. `cgcloud ssh -c example toil-leader`
```
Comment (hannes-ucsc) [inline Production_SOP.md]: Mention briefly that `-c` is same as `--cluster-name`

File: Production_SOP.md [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+2. `virtualenv --system-site-packages toilscripts`
+3. `source toilscripts/bin/activate`
+4. `pip install toil-scripts==2.0.11`
+5. Run the binary for the pipeline, example: `toil-rnaseq` for the RNA-seq CGL pipeline.
+6. Generate a config and manifest file, e.g. `toil-rnaseq generate`. These files will be generated in the current working directory, and contain instructions on how to fill them in.
+7. `exit` so we can grow the cluster, in this case by 22 workers, ideally using spot market.
+8. `cgcloud grow-cluster --cluster-name example --spot-auto-zone --spot-bid 1.00 --instance-type c3.8xlarge --num-workers 22 toil`
+9. `cgcloud ssh -c example toil-leader`
+10. Place boto or AWS credentials on the leader (`~/.boto` or `~/.aws/credentials`)
+10. `screen -S example-run`
```
Comment (hannes-ucsc) [inline Production_SOP.md]: 10 is duplicated

File: Production_SOP.md [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+- `cgcloud create-cluster --leader-instance-type m3.medium --share=key_dir/ --num-workers 0 --cluster-name example toil`
+
+### Setup and Run Pipeline
+SSH into leader, install toil-scripts, setup config and manifest, grow cluster, and start run.
+
+1. `cgcloud ssh -c example toil-leader`
+2. `virtualenv --system-site-packages toilscripts`
+3. `source toilscripts/bin/activate`
+4. `pip install toil-scripts==2.0.11`
+5. Run the binary for the pipeline, example: `toil-rnaseq` for the RNA-seq CGL pipeline.
```
Comment (hannes-ucsc) [inline Production_SOP.md]: `toil-rnaseq`  is not a binary

Comment (hannes-ucsc) [reply Production_SOP.md]: Isn't it "RNASeq"?

File: Production_SOP.md [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+6. Generate a config and manifest file, e.g. `toil-rnaseq generate`. These files will be generated in the current working directory, and contain instructions on how to fill them in.
+7. `exit` so we can grow the cluster, in this case by 22 workers, ideally using spot market.
+8. `cgcloud grow-cluster --cluster-name example --spot-auto-zone --spot-bid 1.00 --instance-type c3.8xlarge --num-workers 22 toil`
```
Comment (hannes-ucsc) [inline Production_SOP.md:44]: Add section on choosing spot market price.

File: Production_SOP.md [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+- `--workDir=/var/lib/toil` - Sets the work directory to the RAID-mounted ephemeral drives
+- `--batchSystem=mesos` - Distributed runs on EC2 use the Mesos batch system
+- `--mesosMaster=mesos-master:5050` - Tells Mesos what port to use
+- `--sseKey` - Encrypts the data in-flight when using the S3 job store for patient security
+- The final positional argument is the location of the jobStore
+
+## Handling Errors
+
+Sometimes a pipeline fails. This can be for various reasons, but it is unlikely that there is an issue with the pipeline
+or Toil itself. Failures generally stem from an issue with the sample -- improperly paired reads, samples too large for
```
Comment (hannes-ucsc) [inline Production_SOP.md]: I like your confidence but there is little evidence for the claim in the second sentence.

File: Production_SOP.md [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+the amount of memory or disk on a machine, etc.
+This assumption isn't based on hubris, but the fact that the pipeline had a throughput of 99.6% across 20,000 samples.
```
Comment (hannes-ucsc) [inline Production_SOP.md]: This sounds arrogant. It was a different version of the pipeline and a different version of Toil. And we had plenty of problems during the run. The fact that Toil+Pipeline were able to run 20k samples with plenty of hand-holding doesn't mean that Joe Shmo should have no problems with their 20 samples.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: Production_SOP.md (hunk 2 above)
Comment (jvivian) [reply Production_SOP.md]: > "periodically" is imprecise. When exactly should they update?

That's a good question, I replaced it with: 

" `toil-box` images will have to be updated following major releases of Toil that contain critical features or bug fixes, or at the direction of the Toil development team."

File: Production_SOP.md (hunk 3 above)
Comment (jvivian) [reply Production_SOP.md]: Corrected.

File: Production_SOP.md (hunk 1 above)
Comment (jvivian) [reply Production_SOP.md:20]: Substituted in your comment, added `~/.profile` options.

File: Production_SOP.md (hunk 4 above)
Comment (jvivian) [reply Production_SOP.md]: I replaced with the full name for clarity.

File: Production_SOP.md (hunk 6 above)
Comment (jvivian) [reply Production_SOP.md]: /s/binary/entrypoint.

Typically I see [RNA-seq](https://en.wikipedia.org/wiki/RNA-Seq)

File: Production_SOP.md (hunk 7 above)
Comment (jvivian) [reply Production_SOP.md:44]: Added section.

File: Production_SOP.md (hunk 5 above)
Comment (jvivian) [reply Production_SOP.md]: Corrected.

File: Production_SOP.md (hunk 8 above)
Comment (jvivian) [reply Production_SOP.md]: I removed this section completely.

File: Production_SOP.md (hunk 9 above)
Comment (jvivian) [reply Production_SOP.md]: I agree, and not appropriate for technical documentation. I just removed this section.

**Comment by jvivian** [comment]
@hannes-ucsc — Ready for review. I believe @JakeNarkizian plans on running the next production batch, so maybe he can use this SOP for that and provide feedback first.

**Comment by nsjake** [comment]
@jvivian I started a production run on monday night already, but I can simulate a production order.

**Review by hannes-ucsc — APPROVED** [review]

**Pull request merged**